File: common/src/main/java/juuxel/adorn/lib/registry/RegistrarFactory.java
package juuxel.adorn.lib.registry;

import juuxel.adorn.util.InlineServices;
import juuxel.adorn.util.Services;
import net.minecraft.core.Registry;
import net.minecraft.network.syncher.EntityDataSerializer;
import net.minecraft.resources.ResourceKey;

@InlineServices
public interface RegistrarFactory {
    <T> KeyedRegistrar<T> create(ResourceKey<Registry<T>> registryKey);

    BlockRegistrar createBlocks();
    ItemRegistrar createItems();

    Registrar<EntityDataSerializer<?>> createForTrackedDataHandlers();

    @InlineServices.Getter
    static RegistrarFactory get() {
        return Services.load(RegistrarFactory.class);
    }
}

File: common/src/main/kotlin/juuxel/adorn/lib/registry/RegistrarFactory.kt
package juuxel.adorn.lib.registry

import juuxel.adorn.util.InlineServices
import juuxel.adorn.util.loadService
import net.minecraft.registry.Registry
import net.minecraft.registry.RegistryKey

interface RegistrarFactory {
    fun <T> create(registryKey: RegistryKey<Registry<T>>): Registrar<T>

    @InlineServices
    companion object {
        private val instance: RegistrarFactory by lazy { loadService() }
        fun get() = instance
    }
}
